### PR TITLE
keepalive : added new function del_destination and added cfg functions

### PR DESCRIPTION
**Pull request by ycaner06**
#### Pre-Submission Checklist

- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [x] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:

- [ ] PR should be backported to stable branches
- [x] Tested changes locally
- [ ] Related to issue #XXXX (replace XXXX with an open issue number)

#### Description
added new function del_destination that deletes sip address from list
added add_destination module functions for kamailio.cfg
added counter parameter for attempt count. after count pass, module dont try until it adds again.
added kemi interfaces both of them
added lock to stack
added find_destination function

### CODE REVIEW

**Comment by ycaner06** [comment]
There is two issue,

- I couldn't decide how to clean old address that passed counter, cleaning with a timer thread or when counter passed limit .

-  When address is a tcp connection , callback cant catch state. how to solve this problem?

After reviews , i will add doc for functions.

**Comment by henningw** [comment]
@gbour - some comments from your side about this extensions from Yasin?

**Comment by henningw** [comment]
It seems that Guilliaume recently changed jobs, see if some others developers can comment here.

**Comment by miconda** [comment]
A quick look and there are some issues, I will post specific details soon.

File: src/modules/keepalive/keepalive_api.c [hunk 1]
```diff
@@ -63,9 +64,15 @@ int ka_add_dest(str *uri, str *owner, int flags, ka_statechanged_f callback,
...
 	struct sip_uri _uri;
-	ka_dest_t *dest;
+	ka_dest_t *dest=0;
 
-	LM_INFO("adding destination: %.*s\n", uri->len, uri->s);
+	LM_DBG("adding destination: %.*s\n", uri->len, uri->s);
+
+	if(ka_find_destination(uri , owner , &dest , &dest)){
+		LM_INFO("uri [%.*s] already in stack --ignoring \r\n",uri->len, uri->s);
+		dest->counter=0;
```
Comment (miconda) [inline src/modules/keepalive/keepalive_api.c]: The change of the `dest->counter` is done out of the mutex zone and this operation can be done on an invalid structure if the `dest` was removed meanwhile.

File: src/modules/keepalive/keepalive_api.c [hunk 2]
```diff
@@ -142,3 +140,100 @@ ka_state ka_destination_state(str *destination)
...
+* @result 1 successful  , -1 fail
+*/
+int ka_del_destination(str *uri, str *owner){
+
+	ka_dest_t *target=0,*head=0;
+
+	if(!ka_find_destination(uri,owner,&target,&head)){
+		LM_ERR("Couldnt find destination \r\n");
+		return -1;
+	}
```
Comment (miconda) [inline src/modules/keepalive/keepalive_api.c]: Similar issues, `target` and `head` are retrieved from `ka_find_destination()`, followed by code out of the mutex zone  and later used directly, but at that time any of them can be already deleted. So removal has to be done in the same mutex zone as it is searched.

Actually I do not see any good use for  `ka_find_destination()` returning `target` and `head`. The function can be useful to know if the address exists, but using `target` and `head` after this function expose to segfault. Either you keep the lock set in the function at the return time and unlock later, out of the function, or just make it to return true/false on finding the destination, without returning `target` and `head`.

File: src/modules/keepalive/keepalive_mod.c [hunk 3]
```diff
@@ -52,19 +52,29 @@ static int ka_mod_add_destination(modparam_t type, void *val);
...
+static int w_add_destination(sip_msg_t *msg, char *uri, char *owner);
+static int w_del_destination(sip_msg_t *msg, char *uri, char *owner);
+
 
 extern struct tm_binds tmb;
 
 int ka_ping_interval = 30;
 ka_destinations_list_t *ka_destinations_list = NULL;
+str ka_ping_from = str_init("sip:keepalive@kamailio.org");
+int counter_del = 5;
```
Comment (miconda) [inline src/modules/keepalive/keepalive_mod.c]: I suggest adding `ka_` prefix to global variables that are not static and exposed in other files via `extern`.

File: src/modules/keepalive/keepalive_mod.c [hunk 4]
```diff
@@ -52,19 +52,29 @@ static int ka_mod_add_destination(modparam_t type, void *val);
...
 
 
 static cmd_export_t cmds[] = {
 	{"is_alive", (cmd_function)w_cmd_is_alive, 1,
 			fixup_spve_null, 0, ANY_ROUTE},
 	// internal API
+	{"add_destination", (cmd_function)w_add_destination, 2,
+		fixup_add_destination, 0, REQUEST_ROUTE|BRANCH_ROUTE|ONREPLY_ROUTE},
+	{"del_destination", (cmd_function)w_del_destination, 2,
+		fixup_add_destination, 0, ANY_ROUTE},
```
Comment (miconda) [inline src/modules/keepalive/keepalive_mod.c]: As more functions are exported from this module, I think is better to also prefix the exported name with `ka_`, like most modules use a common prefix for their functions. Probably we should also add an alias named  `ka_is_alive()` for `is_alive()`.

File: src/modules/keepalive/keepalive_mod.c [hunk 5]
```diff
@@ -126,6 +138,8 @@ static int mod_init(void)
  */
 static void mod_destroy(void)
 {
+	lock_release(ka_destinations_list->lock);
+	lock_dealloc(ka_destinations_list->lock);
 }
```
Comment (miconda) [inline src/modules/keepalive/keepalive_mod.c]: This two have to be enclosed in `if(ka_destinations_list) { ... }`, the mod_destroy() is executed even when mod_init() is not finished, in that case ka_destinations_list is NULL, crashing kamailio at shutdown.

**Comment by ycaner06** [comment]
Thanks for review. I will work on these at weekend.

**Comment by ycaner06** [comment]
if it is ok , will add doc.

**Comment by miconda** [comment]
You can add the docs.

**Comment by miconda** [comment]
Thanks!

**Pull request merged**